gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-common/src/main/java/io/gravitee/am/gateway/handler/common/auth/user/impl/UserAuthenticationManagerImpl.java

```diff
@@ -85,26 +89,15 @@
     public Single<User> authenticate(Client client, Authentication authentication, boolean preAuthenticated) {
         logger.debug("Trying to authenticate [{}]", authentication);
 
-        // Get identity providers associated to a client
-        // For each idp, try to authenticate a user
-        // Try to authenticate while the user can not be authenticated
-        // If user can't be authenticated, send an exception
-
-        // Skip external identity provider for authentication with credentials.
-        List<String> identities = client.getIdentities() != null ?
-                client.getIdentities()
-                        .stream()
-                        .map(idp -> identityProviderManager.getIdentityProvider(idp))
-                        .filter(idp -> idp != null && !idp.isExternal())
-                        .map(IdentityProvider::getId)
-                        .collect(Collectors.toList()) : null;
-        if (identities == null || identities.isEmpty()) {
-            logger.error("No identity provider found for client : " + client.getClientId());
-            return Single.error(new InternalAuthenticationServiceException("No identity provider found for client : " + client.getClientId()));
+        var applicationIdentityProviders = getApplicationIdentityProviders(client, authentication);
+        if (isNull(applicationIdentityProviders) || applicationIdentityProviders.isEmpty()) {
```

Comment:
Is there a reason we're not using a library like apache commons or guava for these kinds of checks? E.g. `CollectionUtils.isEmpty`.

Reply:
I guess we want to limit the use of tied libraries.

Reply:
I thought so. Perhaps we should write our own as I see a lot of these string/collection null checks.